### PR TITLE
Developing more robust persistence layer

### PR DESCRIPTION
**Pull request by rosesyrett**
merged saving/loading of Hkl pickles into one class, which inherits from generic class which can be used with any persistence method (in theory).

### CODE REVIEW

File: src/diffcalc_API/persistence.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            diffcalcObject: HklCalculation = pickle.load(openedFile)
+
+        return diffcalcObject
+
+
+def get_repo() -> HklCalcRepo:
+    return PicklingHklCalcRepo(Path(savePicklesFolder))
+
+
+async def createPickle(pickleFileName: str) -> Path:
```
Comment (garryod) [inline src/diffcalc_API/persistence.py]: This function name could be made more generic i.e. `createStore`

File: src/diffcalc_API/persistence.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    constraints = Constraints()
+    hkl = HklCalculation(UBcalc, constraints)
+
+    repo = get_repo()
+    await repo.save(pickleFileName, hkl)
+
+    return Path(savePicklesFolder, pickleFileName)
+
+
+def deletePickle(pickleFileName: str) -> Path:
```
Comment (garryod) [inline src/diffcalc_API/persistence.py]: This function name could be made more generic i.e. `deleteStore`

File: src/diffcalc_API/persistence.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+from diffcalc.hkl.calc import HklCalculation
+from diffcalc.hkl.constraints import Constraints
+from diffcalc.ub.calc import UBCalculation
+
+from diffcalc_API.config import savePicklesFolder
+from diffcalc_API.errorDefinitions import attempting_to_overwrite, check_file_exists
+
+
+class HklCalcRepo(ABC):
```
Comment (garryod) [inline src/diffcalc_API/persistence.py]: The term "store" may better describe a persistent structure, the term repository implies structure beyond a single item

File: tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+from diffcalc_API.persistence import HklCalcRepo
+
+
+class FakeHklCalcRepo(HklCalcRepo):
+    def __init__(self, useHkl: HklCalculation):
+        self.hkl = useHkl
+
+    async def save(self, name: str, calc: HklCalculation) -> None:
+        return
```
Comment (garryod) [inline tests/conftest.py]: `pass` is typically used in this scenario

File: src/diffcalc_API/persistence.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            pickle.dump(obj=calc, file=stream)
+
+    async def load(self, name: str) -> HklCalculation:
+        file_path = self._root_directory / name
+        check_file_exists(file_path, name)
+
+        with open(file_path, "rb") as openedFile:
+            diffcalcObject: HklCalculation = pickle.load(openedFile)
+
+        return diffcalcObject
```
Comment (garryod) [inline src/diffcalc_API/persistence.py]: I would typically move this to it's own module, allowing for one module per implementation

File: src/diffcalc_API/persistence.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    the state which the API is managing (hkl object).
+    """
+
+    @abstractmethod
+    async def save(self, name: str, calc: HklCalculation) -> None:
+        ...
+
+    @abstractmethod
+    async def load(self, name: str) -> HklCalculation:
+        ...
```
Comment (garryod) [inline src/diffcalc_API/persistence.py]: This may be better as a `Protocol` ([PEP544](https://peps.python.org/pep-0544/))

**Review by garryod — CHANGES_REQUESTED** [review]
Looks good, definitely a nice abstraction, comments are all minor

File: src/diffcalc_API/persistence.py (hunk 2 above)
Comment (callumforrester) [reply src/diffcalc_API/persistence.py]: More generally, the abstraction is leaking as a result of `createPickle` and `deletePickle` still being here. I think we want to add `create` and `delete` methods to the persistence class.

Comment (rosesyrett) [reply src/diffcalc_API/persistence.py]: I've now changed this so my 'HklCalcRepo' has create/delete methods. I think this still makes sense from a database point of view,thanks for the comments and suggestion to do it this way :)

File: src/diffcalc_API/persistence.py (hunk 5 above)
Comment (rosesyrett) [reply src/diffcalc_API/persistence.py]: OK, so would this mean for example making a folder called 'persistence' instead, and having 'HklCalcRepo.py' as a file with the definition of HklCalcRepo, and then having another file 'Pickling.py' with PicklingHklCalcRepo definition?

File: src/diffcalc_API/persistence.py (hunk 6 above)
Comment (rosesyrett) [reply src/diffcalc_API/persistence.py]: I agree - I didn't actually know this was something I could do. Makes things a lot cleaner! Thanks.

File: src/diffcalc_API/persistence.py (hunk 5 above)
Comment (garryod) [reply src/diffcalc_API/persistence.py]: I guess there's three structures I might adopt:

- A: Protocol in super
  ```
  diffcalc_API
  ├ hkl_calc_repo.py <- With HklCalcRepo
  └ hkl_calc_repos
    ├ __init__.py <- Empty
    └ pickling.py <- With PicklingHklCalcRepo
  ```
- B: Protocol in module `__init__`
  ```
  diffcalc_API
  └ hkl_calc_repos
    ├ __init__.py <- With HklCalcRepo
    └ pickling.py <- With PicklingHklCalcRepo
  ```
- C: Protocol in module
  ```
  └ hkl_calc_repos
    ├ __init__.py <- Empty
    ├ protocol.py <- With HklCalcRepo
    └ pickling.py <- With PicklingHklCalcRepo
  ```
I think my preference is for either B or C, with a slight lean towards C. Would be interesting in hearing which @callumforrester finds most pythonic

Comment (callumforrester) [reply src/diffcalc_API/persistence.py]: C gets my vote, along with an `__all__` in `__init__.py` to make the imports nicer

**Pull request merged**